connect-inject/endpoints_controller.go

```diff
@@ -356,18 +362,42 @@ func (r *EndpointsController) createServiceRegistrations(pod corev1.Pod, service
 		parsedIP := net.ParseIP(k8sService.Spec.ClusterIP)
 		if parsedIP != nil {
 			taggedAddresses := make(map[string]api.ServiceAddress)
-			for _, servicePort := range k8sService.Spec.Ports {
-				taggedAddressKey := clusterIPTaggedAddressName
-				if servicePort.Name != "" {
-					taggedAddressKey = fmt.Sprintf("%s-%s", clusterIPTaggedAddressName, servicePort.Name)
-				}
 
-				taggedAddresses[taggedAddressKey] = api.ServiceAddress{
-					Address: k8sService.Spec.ClusterIP,
-					Port:    int(servicePort.Port),
+			// When a service has multiple ports, we need to choose the port that is registered with Consul
+			// and only set that port as the tagged address because Consul currently does not support multiple port
+			// on a single service.
+			var k8sServicePort int32
+			for _, sp := range k8sService.Spec.Ports {
+				// If target port is a name, then we need to find the port value from the pod.
+				if sp.TargetPort.Type == intstr.String {
+					targetPortValue, err := portValue(pod, sp.TargetPort.StrVal)
+					if err != nil {
+						return nil, nil, err
+					}
+
+					// If the targetPortValue is the consulServicePort, then this is the service port we'll use as the tagged address.
+					if targetPortValue == int32(consulServicePort) {
+						k8sServicePort = sp.Port
+					}
+				} else if sp.TargetPort.Type == intstr.Int && sp.TargetPort.IntVal != 0 {
+					// If the target port is a non-zero int, we can compare that port directly with the Consul service port.
+					if sp.TargetPort.IntVal == int32(consulServicePort) {
+						k8sServicePort = sp.Port
+					}
+				} else {
+					// If targetPort is not specified, then the service port is used as the target port,
+					// and we can compare the service port with the Consul service port.
+					if sp.Port == int32(consulServicePort) {
+						k8sServicePort = sp.Port
+					}
```

Comment:
As we expect only a single port to match the `consulServicePort`, can we add a `break` inside the block where `portValue == int32(consulServicePort)`? We potentially dont need to go through the other ports, plus it makes it clear that only a single spec.Port is expected to be a match vs the last one that matches.

Reply:
Hmm I don't think we need a break since it's it will only go into the other if blocks if one of them is false. I.e. if one of the if in the if-else evaluates to `true` then it won't evaluate others. WDYT?

Reply:
Done! great catch!

Reply:
I should have phrased this comment better!